src/Controllers/ActionController.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Sparks\Shield\Controllers;
+
+use CodeIgniter\Controller;
+use CodeIgniter\Exceptions\PageNotFoundException;
+use Sparks\Shield\Authentication\Actions\ActionInterface;
+
+/**
+ * Class ActionController
+ *
+ * A generic controller to handle Authentication Actions.
+ *
+ * @package Sparks\Shield\Controllers
+ */
+class ActionController extends Controller
+{
+	/**
+	 * @var \Sparks\Shield\Authentication\Actions\ActionInterface
+	 */
+	protected $action;
+
+	/**
+	 * Perform an initial check if we have a valid action or not.
+	 *
+	 * @param $method
+	 * @param mixed  ...$params
+	 */
+	public function _remap($method, ...$params)
```

Comment:
This feels like an odd use of _remap(). I assume you did this because it needs to come after filters so the constructor was not an option. Maybe extending initController? Or maybe moving all this into a filter itself? Actions are sort of a grey area between middleware and controller, just brainstorming.

Reply:
It is  a little odd, sure. You're correct about the constructor, and I have always preferred to think of `initController` as a system function and pretend it doesn't exist, but I guess it could be used here.

Reply:
I usually use `initController1 for this kind fo stuff (assuming `__construct()` doesn't work) but it's awkward to have to re-pass the parameters for what is essentially "things you would usually do in the constructor". This feels like a re-appropriation of the remap method but maybe that's better than the other options out there.